File: algorithm/322. Coin Change/cpp/CoinChange.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount+1,-1);
        dp[0] = 0;
        for(int i = 0; i < amount+1; i++){
            if(dp[i] == -1) continue;
            for(int j = 0; j < coins.size(); j++){
                if(i-amount+coins[j] <= 0){  // to deal with overflow
                    if(dp[i+coins[j]] == -1)
                        dp[i+coins[j]] = dp[i]+1;
                    else
                        dp[i+coins[j]] = min(dp[i+coins[j]], dp[i] + 1);
                }
            }
        }
        return dp[amount];
    }
};

File: algorithm/322. Coin Change/java/CoinChange.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount+1];
        for(int i = 1; i < amount+1; i++){
            dp[i] = -1;
        }
        for(int i = 0;i < amount+1; i++){
            if(dp[i] == -1) continue;
            for(int coin: coins){
                if(i-amount+coin <= 0){ // avoid overflow
                    if(dp[i+coin] == -1) dp[i+coin] = 1 + dp[i];
                    else dp[i+coin] = Math.min(dp[i+coin], 1 + dp[i]);     
                }
            }
        }
        return dp[amount];
    }
}
